### PR TITLE
Simplify local docker setup.

### LINKED ISSUE #12722: Getting started for docker single-node cluster should be easier
I walked through the getting started guide in [docker.md](https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/docker.md) and I find it too complicated (assuming it's supposed to be the easiest and recommended way to play with K8s).

Some of the problems I've noticed so far:
* Requirements are hard to verify - we require some kernel configuration which I don't how to check without googling it
* We 3 very long commands which are very hard to parse even using docker reference (they do not fit on 30" screen!)
* Some command outputs are outdated

It's also hard for someone to use kubectl as it doesn't have a context and it's not easy to set it correctly (it's minimum 2 commands which are relatively long)

I think our goal should be that we have a very simple command ```run_local_kubernetes.sh``` that will setup everything (and will verify requirements). Second command should be already starting a pod with a working ```kubectl```.

@brendandburns

### LINKED ISSUE #18227: Containerized hyperkube fails to create certificates
The [`setup`] (https://github.com/kubernetes/kubernetes/blob/d2dbe568205895817b52a71fbef8614f4d9931eb/cluster/images/hyperkube/master.json#L65) container was [recently added](https://github.com/kubernetes/kubernetes/pull/17262) by @brendandburns to automatically generate certificates.

However, [setup-files.sh](https://github.com/kubernetes/kubernetes/blob/ce90b83689f08cb5ebb6b632dab7f95a48060425/cluster/images/hyperkube/setup-files.sh) fails with `chgrp: invalid group: 'kube-cert'`. 

Setting ```CERT_GROUP=`id -g` ``` [when invoking `make-ca-cert.sh`](https://github.com/kubernetes/kubernetes/blob/master/cluster/images/hyperkube/setup-files.sh#L34) fixes it  but after that I get 

```
hostname: Name or service not known
/make-ca-cert.sh: line 21: $1: unbound variable
``` 

Which I have narrowed down to the combination of `hostname -i` with host networking:

```
$ docker run --net=host --rm debian:jessie hostname -i
hostname: Name or service not known
```

```
docker run --rm debian:jessie hostname -i
172.17.0.57
```

`strace` reveals that `hostname -i` in the container simply does a DNS query of the hostname, which won't work in many cases like mine.

```
$ docker run --net=host -ti --rm debian:jessie bash 
root@vagrant-ubuntu-vivid-64:/# hostname
vagrant-ubuntu-vivid-64
root@vagrant-ubuntu-vivid-64:/# apt-get update
[...]
root@vagrant-ubuntu-vivid-64:/# apt-get install strace
[...]
root@vagrant-ubuntu-vivid-64:/# strace hostname -i
[...]
socket(PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 3
connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, 16) = 0
poll([{fd=3, events=POLLOUT}], 1, 0)    = 1 ([{fd=3, revents=POLLOUT}])
sendmmsg(3, {{{msg_name(0)=NULL, msg_iov(1)=[{"\337A\1\0\0\1\0\0\0\0\0\0\27vagrant-ubuntu-vivi"..., 41}], msg_controllen=0, msg_flags=0}, 41}, {{msg_name(0)=NULL, msg_iov(1)=[{"\335\254\1\0\0\1\0\0\0\0\0\0\27vagrant-ubuntu-vivi"..., 41}], msg_controllen=0, msg_flags=0}, 41}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=3, events=POLLIN}], 1, 5000)  = 1 ([{fd=3, revents=POLLIN}])
ioctl(3, FIONREAD, [116])               = 0
recvfrom(3, "\335\254\201\203\0\1\0\0\0\1\0\0\27vagrant-ubuntu-vivi"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, [16]) = 116
poll([{fd=3, events=POLLIN}], 1, 4973)  = 1 ([{fd=3, revents=POLLIN}])
ioctl(3, FIONREAD, [116])               = 0
recvfrom(3, "\337A\201\203\0\1\0\0\0\1\0\0\27vagrant-ubuntu-vivi"..., 1932, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, [16]) = 116
close(3)                                = 0
write(2, "hostname: ", 10hostname: )              = 10
write(2, "Name or service not known", 25Name or service not known) = 25
write(2, "\n", 1
)                       = 1
exit_group(1)                           = ?
+++ exited with 1 +++
```

I would propose:

* Use `hostname -I` (note the capital `I`) and provide a SANs for all the interface IPs.
* Add SANs for the kubernetes service IP and all the full or partial domain names accesible when using the DNS add-on, so that pods can connect to the API server with service accounts. That is `IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local`

I am aware that the `.cluster.local` domain is configurable but it's a reasonable default.

### PR DESCRIPTION
**Pull request by fgrzadkowski**
This simplifies single node cluster setup to just one command.

@brendandburns @dalanlan @roberthbailey @mikedanese 

This partially fixes #12722
Fixes #18227

### CODE REVIEW

**Comment by luxas** [comment]
@fgrzadkowski 
This is quite similar to #17213...
Which one should be merged first?
It has been pending for a while now

**Comment by fgrzadkowski** [comment]
@luxas Good point :) There are some common parts, but most of it is separate. I believe someone will have conflict anyway and will have to merge with the other.

By the way - I've added a comment/question in your PR.

**Comment by luxas** [comment]
@fgrzadkowski Yes, most of the things is separate, however it would be good to merge #17213 so one don't have to do the same patches again and we're ending up with many PRs that share the same code :)

But, how is the `docker-multinode` guide supposed to look in the future? This must harmonize with `bootstrap-docker` usage. In that case, we have to start `etcd` before `flannel` which must be started before docker (I do think you know this already, but just clarifying). So the question is, should we run a `headless` (apiserverless) kubelet for `bootstrap-docker` or how do we do this?

In just `master.json` usage, this is fine, but for `master-multi.json`, we'll have to tweak it further.

BTW, I have run `kube-proxy` as a part of the master static pod for a long time now and it works fine **on master, but not nodes.** So could you `COPY kube-proxy.json /etc/kubernetes/manifests-multi/kube-proxy.json`, too?

File: cluster/images/hyperkube/etcd.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "hostNetwork": true,
+    "containers": [
+      {
+        "name": "etcd",
+        "image": "gcr.io/google_containers/etcd:2.2.1",
+        "command": [
+                "/usr/local/bin/etcd",
+                "--listen-client-urls=http://127.0.0.1:4001",
+                "--advertise-client-urls=http://127.0.0.1:4001",
+                "--data-dir=/var/etcd/data"
```
Comment (brendandburns) [inline cluster/images/hyperkube/etcd.json]: probably want to mount in the data director via emptyDir (I know the old instructions didn't do that...)

**Comment by brendandburns** [comment]
this basically LGTM, but I agree we should get #17213 in too.

**Comment by dalanlan** [comment]
You might want to address #18227

File: cluster/images/hyperkube/kube-proxy.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (roberthbailey) [inline cluster/images/hyperkube/kube-proxy.json]: Woohoo, kube-proxy in a pod!

Comment (dalanlan) [inline cluster/images/hyperkube/kube-proxy.json]: IIRC, it's necessary to pass `--resource-container=""` for kube-proxy in this case. 
Ref [#17619(comment)](https://github.com/kubernetes/kubernetes/issues/17619#issuecomment-160779388)

File: cluster/images/hyperkube/etcd.json (hunk 1 above)
Comment (fgrzadkowski) [inline cluster/images/hyperkube/etcd.json]: Done.

File: cluster/images/hyperkube/kube-proxy.json (hunk 2 above)
Comment (fgrzadkowski) [inline cluster/images/hyperkube/kube-proxy.json]: Done

**Comment by fgrzadkowski** [comment]
@brendandburns Comments addressed. PTAL.

**Comment by luxas** [comment]
@fgrzadkowski What is `--hostname-override` used for? 
Some months ago on my Pis, `hyperkube kubelet` didn't start up correctly without it
But now, I just removed the command and it worked fine.
I just ask because if we don't need to specify this flag, we won't depend on `hostname -i` anymore (`master-multi`)
Could I make a PR that removes it or is it really used for something other than displaying IP instead of the hostname?

**Comment by brendandburns** [comment]
LGTM.
